### PR TITLE
Add generateJob() and GenerateRandomHash

### PR DESCRIPTION
**Pull request by thomashorrobin**
This commit adds functionality to generate a job from a cronjob. To do this locally we need to emulate the code to generate the hashes. Both files have been moved from SwiftkubeClient

### CODE REVIEW

**Comment by iabudiab** [comment]
@thomashorrobin 👍 thx for the PR, I'll take a look tomorrow and report here 😉

File: Sources/Extensions/CronJobExtensions+batch.v1.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	case jobSpecDoesntExist
+	case jobMetadataDoesntExist
+	case cronjobNameDoesntExist
+}
+
+public extension batch.v1.CronJob {
+	func generateJob() throws -> batch.v1.Job {
+		guard let jobTemplateSpec = spec?.jobTemplate.spec else { throw createJobFromCronjobErrors.jobSpecDoesntExist }
+		guard let name = name else { throw createJobFromCronjobErrors.cronjobNameDoesntExist }
+		let jobName = "\(name)-manual-\(GenerateRandomHash(length: 3))"
```
Comment (iabudiab) [inline Sources/Extensions/CronJobExtensions+batch.v1.swift:31]: How providing an optional `name` argument, that defaults to `manual`, i.e. 

```swift
func generateJob(withName name: String = "manual") throws -> batch.v1.Job {
  ...
  guard let cronJobName = .... 
  let jobName = "\(cronJobName)-\(name)-\(GenerateRandomHash(length: 3))"
}
```

**Review by iabudiab — COMMENTED** [review]
You can try to run `swiftformat` in the root folder of the repository. It will pick up the `.swiftfomrat` config file and apply the linting rules.

**Comment by thomashorrobin** [comment]
> You can try to run `swiftformat` in the root folder of the repository. It will pick up the `.swiftfomrat` config file and apply the linting rules.

So I tried that, but it made changes to many files that I hadn't touched, that's why I didn't include it. I have now thought pushed the change: f669e4247914039fbfeaea40a0e18f0dde49e7df

File: Sources/Extensions/CronJobExtensions+batch.v1.swift (hunk 1 above)
Comment (thomashorrobin) [reply Sources/Extensions/CronJobExtensions+batch.v1.swift]: Agreed. Have changed to make "manual" the default name with an optional override 9763391028b6ed5e7ac08afee1fd8a237679e37b

File: Sources/Extensions/Hashes.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	while neededAlphanums > 0 {
+		let nextAlphanum = alphanums.randomElement()!
+		if nextAlphanum != seedAlphanum {
+			seedAlphanum = nextAlphanum
+			slug = slug + nextAlphanum
+			neededAlphanums -= 1
+		}
+	}
+	return slug
+}
```
[the remark this answers is not in the document]
Comment (thomashorrobin) [reply Sources/Extensions/Hashes.swift:39]: Yeah that's cleaner but I've done it this way was to mimic what other Kubernetes tools do. Basically it's set out like that to stop rude words appearing. And the vowels are removed so other valid three character English words don't come up like "cat" or "fox" that might be confused with names for other things

[the remark this answers is not in the document]
Comment (thomashorrobin) [reply Sources/Extensions/Hashes.swift:39]: Let me know if you want me to simplify it though 👍

File: Sources/Extensions/CronJobExtensions+batch.v1.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+import Foundation
+
+// MARK: - createJobFromCronjobErrors
+
+enum createJobFromCronjobErrors: Error {
+	case jobSpecDoesntExist
+	case jobMetadataDoesntExist
+	case cronjobNameDoesntExist
+}
```
[the remark this answers is not in the document]
Comment (thomashorrobin) [reply Sources/Extensions/CronJobExtensions+batch.v1.swift]: Yeah, agreed. I've also made the error more dynamic: 97ab54322aa1323a7642aca464f539b8cf4f2ee2

**Comment by thomashorrobin** [comment]
Hi @iabudiab, this PR is a bit old but still valid as far as I can see. Sorry I almost completely forgot about until I saw a notification about it yesterday!

**Comment by iabudiab** [comment]
@thomashorrobin no worries at all. Many things come in between and have higher priorities.

I'll take a look right away.

File: Sources/Extensions/Hashes.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (iabudiab) [reply Sources/Extensions/Hashes.swift:39]: Sound good. Let's keep it this way.

**Review by iabudiab — APPROVED** [review]

**Pull request merged**